src/io/probst/idea/clangformat/ClangFormatAutoAction.java

```diff
@@ -1,26 +1,66 @@
 package io.probst.idea.clangformat;
 
-import java.util.Arrays;
-import java.util.List;
+import com.intellij.codeInsight.actions.FormatChangedTextUtil;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.codeStyle.ChangedRangesInfo;
+import com.intellij.util.diff.FilesTooBigForDiffException;
+
+import java.util.Collection;
+import java.util.Collections;
 
 /**
  * Runs clang-format on the current selection or on the whole file if there is no selection.
  * Applies the formatting updates to the editor.
  */
 public class ClangFormatAutoAction extends ClangFormatAction {
-    /**
-     * Builds a list of arguments to clang-format
-     */
     @Override
-    protected List<String> getCommandArguments(String clangFormatBinary, String filePath, int cursor, int selectionStart,
-                                               int selectionLength) {
-        if (selectionLength > 0) {
-            // Format the selection as normal
-            return super.getCommandArguments(clangFormatBinary, filePath, cursor, selectionStart, selectionLength);
+    protected Collection<TextRange> getFormatRanges(Project project,
+                                                    Document document,
+                                                    Editor editor,
+                                                    VirtualFile virtualFile) {
+        Settings settings = Settings.get();
+
+        // IntelliJ reports a cursor at the end of the file as being at file length + 1, which breaks
+        // clang-format.
+        int docLength = document.getTextLength() - 1;
+        int selectionStart = Math.min(editor.getSelectionModel().getSelectionStart(), docLength);
+        int selectionLength = Math.min(editor.getSelectionModel().getSelectionEnd(), docLength) - selectionStart;
+
+        // Formatting the VCS changed text will only be done if nothing is selected. Otherwise you can't force a reformat
+        // of some code that was not changed from the VCS version
+        if (selectionLength <= 0 && settings.updateOnlyChangedText) {
+            // This case only formats the text changed in the VCS. If the diff is too large or we are not in a VCS file
```

Comment:
can you extract a method for this branch here? it's conceptually different enough to warrant its own place and name.

Reply:
I extracted the function.